### ISSUE TITLE
Neo Propagation/Translation Method Issues

### ISSUE THREAD

**Issue by riusi88** [spicywebau/craft-neo#355]
**Summary:**
We are currently experiencing an issue where the Translation Method on internal block type fields are not always being respected. A Neo field saved on one site sometimes has it's fields contents propagated to other sites, and overwriting content. The issue does not occur all the time, but it is persistent. 

**Project Overview:**
The project has many language specific multi-sites, with Sections that have a propagation setting set to "Save entries to all sites enabled for this section", and a Neo block set to "Save blocks to all sites the owner element is saved in", with individual fields within the Neo block types set to "Translate for each language".

It's worth noting that there are also many user groups, that only have permissions to edit one or two of the multi-sites.

**The Issue:**
Our users report that sometimes their translations get propagated to other languages they don't have permissions for, and that they were not editing at the time of saving there changes. For example, a user editing in Spanish had their changes propagated to English. What's unfortunate is that when this happens, our users lose their progress in the language they were editing, and we have to roll back our entries to previously saved drafts or versions.

I believe the core issue stems from the individual fields within the Neo block types not respecting the Translation Method -> "Translate for each language" feature, and somehow the fields contents are propagating to the wrong languages.

I'm wondering if somehow multiple users working on the same Entry in different languages might somehow trigger this issue as well? The most common languages experiencing this issue have been es-MX, pt, and they propagate and overwrite en. 

**Info:**
Neo version: 2.7.11
Craft version: 3.4.20
Multi-Site: Yes, 10 multi-sites, each a separate language.
Site languages: en, th, vi, ko, zh-Hans, zh-Hant, ja, de, es-MX, pt
User Groups: 11 user groups, one per language, and one super user. 

Please let me know if you'd want a copy of the database, error logs, and composer.json.

**Comment by riusi88** [comment]
Hi @ttempleton thank you for looking into this. 

Can I request that you take a copy of my db and composer.json/lock files? We've just recently had to stop all edits on the site due to this issue, and won't be starting up again until it's resolved unfortunately.

**Comment by riusi88** [comment]
One of our users just experienced this after attempting to make changes on a Vietnamese version of a page. After logging in they noticed Korean and English content had overwritten her work. Unfortunately I don't know the steps that were taken to reproduce this issue, but perhaps it may be associated with the  "Merge Changes Into Draft" feature in Craft CMS as of 3.4? Furthermore whats also odd is that I believe more then just the Neo fields have been overwritten in some cases, including the Page Title which is starting to lead me to believe that the issue might be bigger then just NEO. 
![image011](https://user-images.githubusercontent.com/5905213/85562282-56be8780-b5fa-11ea-9856-84fa5cc1f80e.jpg)

**Comment by ttempleton** [comment]
Hmm, that is strange.  If you'd still like to send us your database and composer.json/lock, could you please send them to plugins@spicyweb.com.au.  That said, I will have a play around with merging changes into a draft on my local testing setup, and see if I can reproduce this for Neo or Craft in general.

**Comment by riusi88** [comment]
Hi @ttempleton I was actually able to replicate some issues that out users were experiencing, and recorded a  Loom video:

https://www.loom.com/share/05795f13cf754bb68ae6533c478a48e3

**The steps to replicate were:**

1. Save a draft of an Entry in English (or any language). Drafts also get automatically saved in Preview mode. 
2. Open another browser tab and load the same Entry, but this time in a different language (let's say Spanish). Make a small change and Save it. 
3. Now go back to the English version of the Entry, load the Draft that you saved in step one, and publish your changes. You could also optionally click "Merge Changes Into Draft" before publishing changes which produced the same issues. 
4. Refresh the Spanish Entry, and you'll see that you've lost the edits you made in Step #2, or what sometimes happens very rarely is the entire Neo field contents are overwritten with English which I was lucky enough to capture in my loom video. The later issue seems to happen more often when you click  "Merge Changes Into Draft" before publishing the draft changes in Step #3. 

I hope this helps!

**Comment by ttempleton** [comment]
Thanks for the additional information and video, @riusi88, with that I've been able to reproduce the loss of edits in the second site's entry.  Thanks also for sending the email, unfortunately I could not get the database, but I don't think it will be necessary to send that anymore, now that I've reproduced (most of) the issue locally.

The one thing I haven't reproduced yet is the overwriting with data from other sites, but arguably we're overwriting site data in any case (just with older data from the same site in most cases) so I'm hopeful that fixing the general edit loss would fix the overwriting with data from other sites.  I'll keep looking into this and see how we can fix it.

**Comment by riusi88** [comment]
Hi @ttempleton thanks for working on a fix for this. Do you have an ETA for the next release that would include a fix?

**Comment by ttempleton** [comment]
Hi @riusi88,

I'm aiming to have a fix for this completed and released by the end of the month.  Having looked into it further, it seems part of the issue is that draft blocks are separate to the published blocks, with different block IDs, so the merging of published entry data into a draft currently has no way of knowing which blocks to merge into.  This applies to both the "merge changes into draft" button and publishing the draft, because Craft attempts the merge when publishing a draft anyway.

**Comment by DylanEuteneier** [comment]
Any update on this bug? We have a Multilingual site with Neo installed that has experienced this content overwrite issue quite a few times with the use of preview.

**Comment by ttempleton** [comment]
After looking further into this, it seems we unfortunately won't be able to add support for merging published Neo block content into draft blocks at this time.  I believe this is related to craftcms/cms#5503; that issue gives a good explanation of why this happens, which is described in the context of Matrix fields, but clearly also applies to Neo.  That issue doesn't mention multi-site scenarios like this issue describes, but the same explanation still applies as the draft/published content merge still happens for any site where the published content needs to be merged.  I believe that issue would need to be resolved before we can add support for it in Neo.

I'll keep looking into whether there's any way around it that could let us implement this in Neo earlier, but it seems unlikely at this stage.

**Comment by richardfrankza** [comment]
Hi @ttempleton  

We're also experiencing this issue which is causing a lot of frustration for our translators, including overwrite of all translateable content as experienced and demonstrated by @riusi88 in the video.

Are there any suggestions for a workaround at this stage in terms of dos and don'ts for our translators while this gets fixed upstream and hopefully by yourselves eventually.

What can we tell them and is there anything we can disable at a configuration level to minimise the issue? I'm thinking of ways to disable the preview and disabling drafting entirely - nothing could be worse than overwriting content you're not supposed to, so any suggestions we will take onboard.

**Comment by ttempleton** [comment]
@richardfrankza, I don't think there's any official way to disable drafts or previews; I *think* the closest thing to disabling previews would be to remove the preview targets on a section.  But the draft and preview buttons could be hidden, if you're using a plugin like [Control Panel JS](https://plugins.craftcms.com/cp-js) (or if you have another way that you can add custom JavaScript to the control panel), then you could hide the buttons with:

```javascript
$('#preview-btn').addClass('hidden')
$('#save-draft-btn-container').addClass('hidden')
```

**Comment by joeRob2468** [comment]
@richardfrankza Just a temporary fix that might help, in addition to hiding the buttons - looks like they just added a config setting to turn off draft autosaving in the latest release (craftcms/cms#6704). Doesn't address the core issue, but it's a workaround that should work in the meantime.

**Comment by mildlygeeky** [comment]
Does anyone know if this issue is now resolved with 3.7? https://github.com/craftcms/cms/issues/5503#issuecomment-812137674

**Comment by ttempleton** [comment]
Yes, since that Craft enhancement has been resolved for Craft 3.7, we've been able to do what we need to on Neo's end to add compatibility, and [Neo 2.11.0-beta.1](https://github.com/spicywebau/craft-neo/releases/tag/2.11.0-beta.1) has just been released with those changes for anyone running Craft 3.7.0-beta.5.
